File: src/rules/follow-rules.ts
/**
 * Following rules for 升级 trick-taking.
 *
 * When a player leads, they can play anything.
 * When following:
 *   1. Must match the lead pattern type (single→single, pair→pair, etc.)
 *   2. Must follow suit if possible (for the lead suit)
 *   3. If can't follow suit with the required pattern, can play anything
 */

import { Card, Suit, Rank } from '../core/card';
import { PlayType, type PlayAction } from '../core/types';
import { effectiveSuit, trumpOrder } from './comparator';

/**
 * Determine the lead suit of a trick.
 * For the first card played, its effective suit is the lead suit.
 * For trump leads, the lead suit is 'trump'.
 */
export function getLeadSuit(leadAction: PlayAction, trumpSuit: Suit, trumpRank: Rank): Suit | 'trump' {
  if (leadAction.cards.length === 0) return Suit.JOKER;
  return effectiveSuit(leadAction.cards[0], trumpSuit, trumpRank);
}

/**
 * Check if a player CAN follow the lead pattern with their hand.
 *
 * @returns true if the player has cards that can follow the lead.
 */
export function canFollow(
  hand: Card[],
  leadAction: PlayAction,
  trumpSuit: Suit,
  trumpRank: Rank,
): boolean {
  switch (leadAction.type) {
    case PlayType.SINGLE:
      return canFollowSingle(hand, leadAction.cards[0], trumpSuit, trumpRank);
    case PlayType.PAIR:
      return canFollowPair(hand, leadAction.cards[0], trumpSuit, trumpRank);
    case PlayType.TRACTOR:
      return canFollowTractor(hand, leadAction, trumpSuit, trumpRank);
    case PlayType.THROW:
      return canFollowThrow(hand, leadAction, trumpSuit, trumpRank);
  }
}

/**
 * Get all legal follow plays for a given lead.
 * Returns an array of all valid card combinations the player can play.
 */
export function getLegalFollows(
  hand: Card[],
  leadAction: PlayAction,
  trumpSuit: Suit,
  trumpRank: Rank,
): PlayAction[] {
  switch (leadAction.type) {
    case PlayType.SINGLE:
      return getFollowSingles(hand, leadAction.cards[0], trumpSuit, trumpRank);
    case PlayType.PAIR:
      return getFollowPairs(hand, leadAction.cards[0], trumpSuit, trumpRank);
    case PlayType.TRACTOR:
      return getFollowTractors(hand, leadAction, trumpSuit, trumpRank);
    case PlayType.THROW:
      return getFollowThrows(hand, leadAction, trumpSuit, trumpRank);
  }
}

// ---- Single following ----

function canFollowSingle(hand: Card[], leadCard: Card, trumpSuit: Suit, trumpRank: Rank): boolean {
  const effSuit = effectiveSuit(leadCard, trumpSuit, trumpRank);

  if (effSuit === 'trump') {
    // Must follow with trump — any trump card works
    return hand.some(c => effectiveSuit(c, trumpSuit, trumpRank) === 'trump');
  }

  // Must follow with same suit (non-trump)
  return hand.some(c =>
    effectiveSuit(c, trumpSuit, trumpRank) === effSuit &&
    c.rank !== trumpRank
  );
}

function getFollowSingles(hand: Card[], leadCard: Card, trumpSuit: Suit, trumpRank: Rank): PlayAction[] {
  const effSuit = effectiveSuit(leadCard, trumpSuit, trumpRank);
  const canFollow = canFollowSingle(hand, leadCard, trumpSuit, trumpRank);

  if (canFollow) {
    // Must follow with matching suit
    return hand
      .filter(c => {
        if (effSuit === 'trump') {
          return effectiveSuit(c, trumpSuit, trumpRank) === 'trump';
        }
        return effectiveSuit(c, trumpSuit, trumpRank) === effSuit && c.rank !== trumpRank;
      })
      .map(c => ({ type: PlayType.SINGLE, cards: [c] }));
  }

  // Can't follow suit — can play any single
  return hand.map(c => ({ type: PlayType.SINGLE, cards: [c] }));
}

// ---- Pair following ----

function canFollowPair(hand: Card[], leadCard: Card, trumpSuit: Suit, trumpRank: Rank): boolean {
  const effSuit = effectiveSuit(leadCard, trumpSuit, trumpRank);
  return hasPairInEffectiveSuit(hand, effSuit, trumpSuit, trumpRank);
}

function getFollowPairs(hand: Card[], leadCard: Card, trumpSuit: Suit, trumpRank: Rank): PlayAction[] {
  const effSuit = effectiveSuit(leadCard, trumpSuit, trumpRank);
  const pairs = findPairsInEffectiveSuit(hand, effSuit, trumpSuit, trumpRank);

  if (pairs.length > 0) {
    // Must follow with a pair of matching suit
    return pairs.map(cards => ({ type: PlayType.PAIR, cards: cards.slice(0, 2) }));
  }

  // Can't follow — can play any pair or any two singles (treated as discards)
  // Actually, if you can't follow a pair, you play any two cards
  return generateAnyTwoCards(hand);
}

// ---- Tractor following ----

function canFollowTractor(hand: Card[], leadAction: PlayAction, trumpSuit: Suit, trumpRank: Rank): boolean {
  const effSuit = effectiveSuit(leadAction.cards[0], trumpSuit, trumpRank);
  const pairCount = leadAction.cards.length / 2;
  return hasTractorInEffectiveSuit(hand, effSuit, pairCount, trumpSuit, trumpRank);
}

function getFollowTractors(hand: Card[], leadAction: PlayAction, trumpSuit: Suit, trumpRank: Rank): PlayAction[] {
  const effSuit = effectiveSuit(leadAction.cards[0], trumpSuit, trumpRank);
  const pairCount = leadAction.cards.length / 2;
  const tractors = findTractorsInEffectiveSuit(hand, effSuit, pairCount, trumpSuit, trumpRank);

  if (tractors.length > 0) {
    // Must follow with a tractor of same suit and length
    return tractors.map(cards => ({
      type: PlayType.TRACTOR,
      cards: cards.slice(0, pairCount * 2),
    }));
  }

  // Can't follow tractor — can play any (pairCount * 2) cards
  // If can follow with some pairs (but not consecutive), play those pairs
  const pairs = findPairsInEffectiveSuit(hand, effSuit, trumpSuit, trumpRank);
  const followCards: Card[] = [];
  for (const pair of pairs) {
    followCards.push(...pair.slice(0, 2));
  }
  // Fill remaining with any cards from same suit or discards
  if (followCards.length >= pairCount * 2) {
    return [{
      type: PlayType.TRACTOR,
      cards: followCards.slice(0, pairCount * 2),
    }];
  }
  // Pad with any cards
  const remaining = hand.filter(c => !followCards.includes(c));
  const all = [...followCards, ...remaining];
  if (all.length >= pairCount * 2) {
    return [{
      type: PlayType.TRACTOR,
      cards: all.slice(0, pairCount * 2),
    }];
  }
  return [];
}

// ---- Throw following ----

function canFollowThrow(hand: Card[], leadAction: PlayAction, trumpSuit: Suit, trumpRank: Rank): boolean {
  const suit = leadAction.cards[0].suit;
  if (suit === Suit.JOKER) return false;

  // Must follow with cards of the same suit
  const matchingHand = hand.filter(c => c.suit === suit && c.rank !== trumpRank);
  return matchingHand.length >= leadAction.cards.length;
}

function getFollowThrows(hand: Card[], leadAction: PlayAction, trumpSuit: Suit, trumpRank: Rank): PlayAction[] {
  const suit = leadAction.cards[0].suit;
  if (suit === Suit.JOKER) return [];

  const matchingHand = hand.filter(c => c.suit === suit && c.rank !== trumpRank);
  const required = leadAction.cards.length;

  if (matchingHand.length >= required) {
    // Must follow with `required` cards of the same suit
    // Generate combinations (simplified: take highest `required` cards)
    return [{
      type: PlayType.THROW,
      cards: matchingHand.slice(0, required),
    }];
  }

  // Can't fully follow — can discard any `required` cards
  // But we must play all matching cards first
  const discards = hand
    .filter(c => !matchingHand.includes(c))
    .slice(0, required - matchingHand.length);
  return [{
    type: PlayType.THROW,
    cards: [...matchingHand, ...discards],
  }];
}

// ---- Helper functions ----

/**
 * Check if there's at least one pair in the given effective suit.
 */
function hasPairInEffectiveSuit(
  hand: Card[],
  effSuit: Suit | 'trump',
  trumpSuit: Suit,
  trumpRank: Rank,
): boolean {
  const group = groupByEffectiveOrder(hand, effSuit, trumpSuit, trumpRank);
  for (const cards of group.values()) {
    if (cards.length >= 2) return true;
  }
  return false;
}

/**
 * Find all pairs in the given effective suit.
 */
function findPairsInEffectiveSuit(
  hand: Card[],
  effSuit: Suit | 'trump',
  trumpSuit: Suit,
  trumpRank: Rank,
): Card[][] {
  const group = groupByEffectiveOrder(hand, effSuit, trumpSuit, trumpRank);
  const pairs: Card[][] = [];
  for (const cards of group.values()) {
    if (cards.length >= 2) {
      pairs.push(cards.slice(0, 2));
    }
  }
  return pairs;
}

/**
 * Check if there's a tractor of given length in the effective suit.
 */
function hasTractorInEffectiveSuit(
  hand: Card[],
  effSuit: Suit | 'trump',
  pairCount: number,
  trumpSuit: Suit,
  trumpRank: Rank,
): boolean {
  const pairs = findConsecutivePairsInGroup(hand, effSuit, trumpSuit, trumpRank);
  return findConsecutiveRun(pairs, pairCount) !== null;
}

/**
 * Find all tractors of given length in the effective suit.
 */
function findTractorsInEffectiveSuit(
  hand: Card[],
  effSuit: Suit | 'trump',
  pairCount: number,
  trumpSuit: Suit,
  trumpRank: Rank,
): Card[][] {
  const pairs = findConsecutivePairsInGroup(hand, effSuit, trumpSuit, trumpRank);
  if (pairs.length < pairCount) return [];

  const results: Card[][] = [];
  for (let i = 0; i <= pairs.length - pairCount; i++) {
    let isConsecutive = true;
    for (let j = i + 1; j < i + pairCount; j++) {
      if (pairs[j].order !== pairs[j - 1].order - getTrumpStep(pairs[j - 1].order, effSuit)) {
        isConsecutive = false;
        break;
      }
    }
    if (isConsecutive) {
      const cards: Card[] = [];
      for (let j = i; j < i + pairCount; j++) {
        cards.push(...pairs[j].cards.slice(0, 2));
      }
      results.push(cards);
    }
  }

  return results;
}

interface OrderGroup {
  order: number;
  cards: Card[];
}

function findConsecutivePairsInGroup(
  hand: Card[],
  effSuit: Suit | 'trump',
  trumpSuit: Suit,
  trumpRank: Rank,
): OrderGroup[] {
  const group = groupByEffectiveOrder(hand, effSuit, trumpSuit, trumpRank);

  // Filter to orders that have pairs, sorted by order descending
  const pairs: OrderGroup[] = [];
  for (const [orderStr, cards] of group) {
    const order = parseInt(orderStr);
    if (cards.length >= 2) {
      pairs.push({ order, cards });
    }
  }
  pairs.sort((a, b) => b.order - a.order);
  return pairs;
}

function findConsecutiveRun(pairs: OrderGroup[], length: number): OrderGroup[] | null {
  if (pairs.length < length) return null;
  for (let i = 0; i <= pairs.length - length; i++) {
    let ok = true;
    for (let j = i + 1; j < i + length; j++) {
      const step = pairs[j - 1].order >= 70 ? 1 : 1; // Simplified
      if (pairs[j].order !== pairs[j - 1].order - step) {
        ok = false;
        break;
      }
    }
    if (ok) return pairs.slice(i, i + length);
  }
  return null;
}

/**
 * Group cards by their effective order in the given suit context.
 */
function groupByEffectiveOrder(
  hand: Card[],
  effSuit: Suit | 'trump',
  trumpSuit: Suit,
  trumpRank: Rank,
): Map<string, Card[]> {
  const map = new Map<string, Card[]>();

  for (const c of hand) {
    const eff = effectiveSuit(c, trumpSuit, trumpRank);
    if (eff !== effSuit) continue;

    const order = trumpOrder(c, trumpSuit, trumpRank);
    const key = String(order);
    const arr = map.get(key) ?? [];
    arr.push(c);
    map.set(key, arr);
  }

  return map;
}

function getTrumpStep(order: number, effSuit: Suit | 'trump'): number {
  if (effSuit === 'trump') {
    if (order === 100) return 10;
    if (order === 90) return 10;
    return 1;
  }
  return 1;
}

/**
 * Generate all ways to play any two cards (for when you can't follow a pair).
 */
function generateAnyTwoCards(hand: Card[]): PlayAction[] {
  if (hand.length < 2) return [];
  // Return one option: play the two lowest cards
  return [{
    type: PlayType.PAIR,
    cards: [hand[hand.length - 2], hand[hand.length - 1]],
  }];
}

File: server/rules/follow_rules.py
"""Following rules for 升级 (Shengji/Tractor) trick-taking.

When a player leads, they can play anything.
When following:
  1. Must match the lead pattern type (single->single, pair->pair, etc.)
  2. Must follow suit if possible (for the lead suit)
  3. If can't follow suit with the required pattern, can play anything

Ported from src/rules/follow-rules.ts.
"""

from __future__ import annotations

from server.engine.card import Card, Suit, Rank
from server.engine.types import PlayType, PlayAction
from server.rules.comparator import effective_suit, trump_order


# ---- Public API ----


def get_lead_suit(
    lead_action: PlayAction, trump_suit: Suit, trump_rank: Rank,
) -> Suit | str:
    """Determine the lead suit of a trick.

    For trump leads, the lead suit is 'trump'.
    For non-trump leads, it is the effective suit of the first card.
    """
    if not lead_action.cards:
        return Suit.JOKER
    return effective_suit(lead_action.cards[0], trump_suit, trump_rank)


def can_follow(
    hand: list[Card],
    lead_action: PlayAction,
    trump_suit: Suit,
    trump_rank: Rank,
) -> bool:
    """Check if a player CAN follow the lead pattern with their hand.

    Returns True if the player has cards that can follow the lead.
    """
    match lead_action.type:
        case PlayType.SINGLE:
            return _can_follow_single(hand, lead_action.cards[0], trump_suit, trump_rank)
        case PlayType.PAIR:
            return _can_follow_pair(hand, lead_action.cards[0], trump_suit, trump_rank)
        case PlayType.TRACTOR:
            return _can_follow_tractor(hand, lead_action, trump_suit, trump_rank)
        case PlayType.THROW:
            return _can_follow_throw(hand, lead_action, trump_suit, trump_rank)
    return False


def get_legal_follows(
    hand: list[Card],
    lead_action: PlayAction,
    trump_suit: Suit,
    trump_rank: Rank,
) -> list[PlayAction]:
    """Get all legal follow plays for a given lead.

    Returns a list of all valid card combinations the player can play.
    """
    match lead_action.type:
        case PlayType.SINGLE:
            return _get_follow_singles(hand, lead_action.cards[0], trump_suit, trump_rank)
        case PlayType.PAIR:
            return _get_follow_pairs(hand, lead_action.cards[0], trump_suit, trump_rank)
        case PlayType.TRACTOR:
            return _get_follow_tractors(hand, lead_action, trump_suit, trump_rank)
        case PlayType.THROW:
            return _get_follow_throws(hand, lead_action, trump_suit, trump_rank)
    return []


# ---- Single following ----


def _can_follow_single(
    hand: list[Card], lead_card: Card, trump_suit: Suit, trump_rank: Rank,
) -> bool:
    eff_suit = effective_suit(lead_card, trump_suit, trump_rank)
    if eff_suit == "trump":
        return any(effective_suit(c, trump_suit, trump_rank) == "trump" for c in hand)
    return any(
        effective_suit(c, trump_suit, trump_rank) == eff_suit and c.rank != trump_rank
        for c in hand
    )


def _get_follow_singles(
    hand: list[Card], lead_card: Card, trump_suit: Suit, trump_rank: Rank,
) -> list[PlayAction]:
    eff_suit = effective_suit(lead_card, trump_suit, trump_rank)
    if _can_follow_single(hand, lead_card, trump_suit, trump_rank):
        matching = [
            c for c in hand
            if effective_suit(c, trump_suit, trump_rank) == (
                "trump" if eff_suit == "trump" else eff_suit
            ) and (eff_suit == "trump" or c.rank != trump_rank)
        ]
        return [PlayAction(type=PlayType.SINGLE, cards=[c]) for c in matching]
    return [PlayAction(type=PlayType.SINGLE, cards=[c]) for c in hand]


# ---- Pair following ----


def _can_follow_pair(
    hand: list[Card], lead_card: Card, trump_suit: Suit, trump_rank: Rank,
) -> bool:
    eff_suit = effective_suit(lead_card, trump_suit, trump_rank)
    return _has_pair_in_effective_suit(hand, eff_suit, trump_suit, trump_rank)


def _get_follow_pairs(
    hand: list[Card], lead_card: Card, trump_suit: Suit, trump_rank: Rank,
) -> list[PlayAction]:
    eff_suit = effective_suit(lead_card, trump_suit, trump_rank)
    pairs = _find_pairs_in_effective_suit(hand, eff_suit, trump_suit, trump_rank)
    if pairs:
        return [PlayAction(type=PlayType.PAIR, cards=pair[:2]) for pair in pairs]
    # Can't follow -- play any two cards
    return _generate_any_two(hand)


# ---- Tractor following ----


def _can_follow_tractor(
    hand: list[Card], lead_action: PlayAction, trump_suit: Suit, trump_rank: Rank,
) -> bool:
    eff_suit = effective_suit(lead_action.cards[0], trump_suit, trump_rank)
    pair_count = len(lead_action.cards) // 2
    return _has_tractor_in_effective_suit(hand, eff_suit, pair_count, trump_suit, trump_rank)


def _get_follow_tractors(
    hand: list[Card], lead_action: PlayAction, trump_suit: Suit, trump_rank: Rank,
) -> list[PlayAction]:
    eff_suit = effective_suit(lead_action.cards[0], trump_suit, trump_rank)
    pair_count = len(lead_action.cards) // 2
    tractors = _find_tractors_in_effective_suit(hand, eff_suit, pair_count, trump_suit, trump_rank)
    if tractors:
        return [
            PlayAction(type=PlayType.TRACTOR, cards=cards[: pair_count * 2])
            for cards in tractors
        ]
    # Can't follow tractor -- try pairs, then fill
    pairs = _find_pairs_in_effective_suit(hand, eff_suit, trump_suit, trump_rank)
    follow_cards: list[Card] = []
    for pair in pairs:
        follow_cards.extend(pair[:2])
    needed = pair_count * 2
    if len(follow_cards) >= needed:
        return [PlayAction(type=PlayType.TRACTOR, cards=follow_cards[:needed])]
    remaining = [c for c in hand if c not in follow_cards]
    all_cards = follow_cards + remaining
    if len(all_cards) >= needed:
        return [PlayAction(type=PlayType.TRACTOR, cards=all_cards[:needed])]
    return []


# ---- Throw following ----


def _can_follow_throw(
    hand: list[Card], lead_action: PlayAction, trump_suit: Suit, trump_rank: Rank,
) -> bool:
    suit = lead_action.cards[0].suit
    if suit == Suit.JOKER:
        return False
    matching = [c for c in hand if c.suit == suit and c.rank != trump_rank]
    return len(matching) >= len(lead_action.cards)


def _get_follow_throws(
    hand: list[Card], lead_action: PlayAction, trump_suit: Suit, trump_rank: Rank,
) -> list[PlayAction]:
    suit = lead_action.cards[0].suit
    if suit == Suit.JOKER:
        return []
    matching = [c for c in hand if c.suit == suit and c.rank != trump_rank]
    required = len(lead_action.cards)
    if len(matching) >= required:
        return [PlayAction(type=PlayType.THROW, cards=matching[:required])]
    # Not enough matching -- play all matching + fill with others
    discards = [c for c in hand if c not in matching][: required - len(matching)]
    return [PlayAction(type=PlayType.THROW, cards=matching + discards)]


# ---- Helper functions ----


def _has_pair_in_effective_suit(
    hand: list[Card], eff_suit: Suit | str, trump_suit: Suit, trump_rank: Rank,
) -> bool:
    group = _group_by_effective_order(hand, eff_suit, trump_suit, trump_rank)
    return any(len(cards) >= 2 for cards in group.values())


def _find_pairs_in_effective_suit(
    hand: list[Card], eff_suit: Suit | str, trump_suit: Suit, trump_rank: Rank,
) -> list[list[Card]]:
    group = _group_by_effective_order(hand, eff_suit, trump_suit, trump_rank)
    return [cards[:2] for cards in group.values() if len(cards) >= 2]


def _has_tractor_in_effective_suit(
    hand: list[Card], eff_suit: Suit | str, pair_count: int,
    trump_suit: Suit, trump_rank: Rank,
) -> bool:
    pairs = _find_consecutive_pairs(hand, eff_suit, trump_suit, trump_rank)
    return _find_consecutive_run(pairs, pair_count, eff_suit) is not None


def _find_tractors_in_effective_suit(
    hand: list[Card], eff_suit: Suit | str, pair_count: int,
    trump_suit: Suit, trump_rank: Rank,
) -> list[list[Card]]:
    pairs = _find_consecutive_pairs(hand, eff_suit, trump_suit, trump_rank)
    if len(pairs) < pair_count:
        return []
    results: list[list[Card]] = []
    for i in range(len(pairs) - pair_count + 1):
        is_consecutive = True
        for j in range(i + 1, i + pair_count):
            step = _get_trump_step(pairs[j - 1][0], eff_suit)
            if pairs[j][0] != pairs[j - 1][0] - step:
                is_consecutive = False
                break
        if is_consecutive:
            cards: list[Card] = []
            for j in range(i, i + pair_count):
                cards.extend(pairs[j][1][:2])
            results.append(cards)
    return results


def _find_consecutive_pairs(
    hand: list[Card], eff_suit: Suit | str, trump_suit: Suit, trump_rank: Rank,
) -> list[tuple[int, list[Card]]]:
    """Find pairs in effective suit, sorted by order descending.

    Returns list of (order, cards) tuples where cards has length >= 2.
    """
    group = _group_by_effective_order(hand, eff_suit, trump_suit, trump_rank)
    pairs = [(order, cards) for order, cards in group.items() if len(cards) >= 2]
    pairs.sort(key=lambda x: x[0], reverse=True)
    return pairs


def _find_consecutive_run(
    pairs: list[tuple[int, list[Card]]], length: int,
    eff_suit: Suit | str = Suit.HEARTS,
) -> list[tuple[int, list[Card]]] | None:
    if len(pairs) < length:
        return None
    for i in range(len(pairs) - length + 1):
        ok = True
        for j in range(i + 1, i + length):
            step = _get_trump_step(pairs[j - 1][0], eff_suit)
            if pairs[j][0] != pairs[j - 1][0] - step:
                ok = False
                break
        if ok:
            return pairs[i : i + length]
    return None


def _group_by_effective_order(
    hand: list[Card], eff_suit: Suit | str, trump_suit: Suit, trump_rank: Rank,
) -> dict[int, list[Card]]:
    """Group cards by their effective order in the given suit context."""
    result: dict[int, list[Card]] = {}
    for c in hand:
        if effective_suit(c, trump_suit, trump_rank) != eff_suit:
            continue
        order = trump_order(c, trump_suit, trump_rank)
        result.setdefault(order, []).append(c)
    return result


def _get_trump_step(order: int, eff_suit: Suit | str) -> int:
    """Get the step size to the next lower trump order level."""
    if eff_suit == "trump":
        if order >= 70:
            return 10  # BJ(100) -> SJ(90) -> trump-rank(80,70-73) all step=10
        return 1
    return 1


def _generate_any_two(hand: list[Card]) -> list[PlayAction]:
    """Generate a fallback play of any two cards (for when you can't follow a pair)."""
    if len(hand) < 2:
        return []
    return [PlayAction(type=PlayType.PAIR, cards=[hand[-2], hand[-1]])]
